pkg/identity/legacy/issuer.go

```diff
@@ -59,14 +60,60 @@ type FulcioConfig struct {
 	lru *lru.TwoQueueCache
 }
 
+func (fc *legacyIssuer) Match(ctx context.Context, url string) bool {
```

Comment:
Additionally, we'd be duplicating `Authenticate`'s implementation for each `Issuer`, correct?

Maybe could we first focus on splitting up the embedding logic into each principal? And after that, see what it would take to add a new issuer?

Reply:
That said, the part of the FulcioConfig that simply gave us the format for the config file should probably still be a public data structure I think...

Reply:
In the original design doc, it noted 4 places that needed to be updated to add a new type - The issuer config string, the mapping between issuer config string and challenge, the challenge, and custom embedding logic.

With an issuer pool design, correct me if I'm wrong, but we'll need the same number of changes - An issuer config string, an initializer that creates issuers for each config string, the authentication/challenge logic, and custom embedding. Then it's just a question of which is more readable - They're pretty similar in my opinion.

Reply:
I think this model works well for something that is expecting external plugins and they need to register themselves. But I don't know if that makes sense here where the fulcio code is fully self-contained with all issuers defined in this codebase. I would expect a single place to direct these decisions to be easier to read and debug.

If the selector is already/would be a simple switch statement, that would be preferable and I think far more readable.

Reply:
Hmm ok these seems to be too big to review so I'll close this and break up the work, but I think we should review what we consider a good end state here.

Personally, I believe that pluggable issuer implementations is the right end goal. My inspiration for this is https://github.com/distribution/distribution. With a handful of lines of code you can recompile the registry with only the storage providers you want and e.g a custom authorization implementation to match your needs.

The [proposal for this refactor](https://docs.google.com/document/d/1PKAou7wEmEob4VOLq6rFlx5maOM5hkJ3XffZizfwdFw/edit) pushes us in that direction by making a clear interface for each OIDC issuer implementation. We already have a clear interface for the other dependencies (a CA backend and CT log client).

@haydentherapper I agree if we keep the current configuration schema an implementation would require changing something in the config code as well, but I think we can improve that. The model used in distribution/distribution for example is simply to look at the type (of storage for example) and just pass that part back to the matching driver. So no driver specific details end up in the global config structure.

If folks feel strongly that this isn't the right approach I think to draw a really clear picture of what that looks like so I don't keep pushing this in the wrong direction

Reply:
It seems like there are a few goals to this refactor: Split up the X.509 extension logic, make it simpler to add new issuers, and provide a pluggable interface.

I like the Principal structure - It's well-named and I think the refactor is increasing code clarity and simplifying the ability to add tests. I'd like to see that refactor wrapped up and see where that leaves us.

As I mentioned above, after the Principal refactor, it's pretty simple to add a new issuer. I'm not concerned about this, so I don't think we need to refactor more to improve adding issuers.
 
I don't think it's a significant issue to compile the service with all of the issuers. In contrast, with something like Cosign and its providers, some of those providers pull in huge libraries like the GCP libraries, so there's a benefit to the size of the binary if providers are made pluggable. At a glance, it seems like distribution/distribution is the same - Large OCI libraries per registry.\
In Fulcio's case, binary size isn't a concern, as the issuers are just small bits of authentication logic. I don't have a concern with Fulcio always supporting all authorization providers, and leaving the control of which of those providers are active to the configuration. Fulcio also is a service, and not a library, so there's not a concern of pulling in unnecessary dependencies to other libraries.\
From a purely code perspective, I'm concerned that code clarity is reduced as we move away from a simple switch statement. The refactor also does not increase the ability to add tests, as it's currently possible to add tests for each challenge type.

Reply:
I don't think I understand what you mean by the Principal refactor. Can you clarify what that means? What would that refactor look like when done?

Reply:
Splitting up [`Embed`](https://github.com/sigstore/fulcio/blob/e1fe2f85ba4dc3277f1c4585e9d3049f72246ee3/pkg/challenges/challenges.go#L81) into implementations of [`Principal`](https://github.com/sigstore/fulcio/blob/main/pkg/identity/principal.go)

When the refactor is done, there will be a Principal implementation per challenge type.

To add a new issuer, you would need to:
* Add a Principal implementation to handle embedding claims in the certificate
* Add a challenge like `username`/`email`/`spiffe` that returns the Principal
* Add a new type to the [config](https://github.com/sigstore/fulcio/blob/e1fe2f85ba4dc3277f1c4585e9d3049f72246ee3/pkg/config/config.go#L183)
* Update the [switch](https://github.com/sigstore/fulcio/blob/e1fe2f85ba4dc3277f1c4585e9d3049f72246ee3/pkg/challenges/challenges.go#L412) statement to add the challenge

Reply:
Hmm ok. I'll do that bit first and argument my point about the pluggable issuers once I'm done.